prov/efa/src/efa_cq.c

```diff
@@ -331,11 +359,210 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
 err_destroy_pool:
 	ofi_bufpool_destroy(cq->wce_pool);
 err_destroy_cq:
-	ibv_destroy_cq(cq->ibv_cq);
+	ibv_destroy_cq(ibv_cq_ex_to_cq(cq->ibv_cq_ex));
 err_free_util_cq:
 	ofi_cq_cleanup(&cq->util_cq);
 err_free_cq:
 	free(cq);
 	return ret;
 }
 
+
+#if EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+#include "efa_unit_test_utils.h"
+
+struct efa_cq_readfrom_state {
+	struct efa_cq *efa_cq;
+	fi_addr_t *src_addr;
+};
+
+struct efa_cq_readerr_state {
+	struct fi_cq_err_entry *entry;
+	struct efa_cq *efa_cq;
+};
+
+static int setup_readfrom(struct efa_cq_readfrom_state *state)
+{
+	int err;
+	static struct efa_cq *efa_cq;
+
+	efa_cq = calloc(1, sizeof(struct efa_cq));
+	if (!efa_cq)
+		return -1;
+
+	err = efa_cq_initialize(efa_cq);
+	if (err)
+		return -1;
+
+	efa_cq->ibv_cq_ex->start_poll = &_start_poll;
+	efa_cq->ibv_cq_ex->next_poll = &_next_poll;
+	efa_cq->ibv_cq_ex->end_poll = &_end_poll;
+	efa_cq->ibv_cq_ex->read_vendor_err = &_read_vendor_err;
+	efa_cq->ibv_cq_ex->read_wc_flags = &_read_wc_flags;
+	efa_cq->ibv_cq_ex->read_qp_num = &_read_qp_num;
+	efa_cq->ibv_cq_ex->read_byte_len = &_read_byte_len;
+	efa_cq->ibv_cq_ex->read_src_qp = &_read_src_qp;
+	efa_cq->ibv_cq_ex->read_sl = &_read_sl;
+	efa_cq->ibv_cq_ex->read_slid = &_read_slid;
+	efa_cq->ibv_cq_ex->read_imm_data = &_read_imm_data;
+	efa_cq->ibv_cq_ex->read_opcode = &_read_send_code;
+	efa_cq->read_entry = &_read_entry;
+
+	state->efa_cq = efa_cq;
+	state->src_addr = NULL;
+
+	return 0;
+}
+
+static int teardown_readfrom(struct efa_cq_readfrom_state *state)
+{
+	teardown_efa_cq(state->efa_cq);
+	free(state->efa_cq);
+	free(state);
+
+	return 0;
+}
+
+static int setup_readerr(struct efa_cq_readerr_state *state)
+{
+	int err;
+	static struct efa_cq *efa_cq;
+	static struct fi_cq_err_entry *fi_cq_err_entry;
+
+	efa_cq = calloc(1, sizeof(struct efa_cq));
+	fi_cq_err_entry = calloc(1, sizeof(struct fi_cq_err_entry));
+
+	if (!efa_cq || !fi_cq_err_entry)
+		return -1;
+
+	err = efa_cq_initialize(efa_cq);
+	if (err)
+		return -1;
+
+	efa_cq->ibv_cq_ex->read_opcode = &_read_send_code;
+
+	state->entry = fi_cq_err_entry;
+	state->efa_cq = efa_cq;
+
+	return 0;
+}
+
+static int teardown_readerr(struct efa_cq_readerr_state *state)
+{
+	teardown_efa_cq(state->efa_cq);
+	free(state->efa_cq);
+	free(state->entry);
+	free(state);
+
+	return 0;
+}
+
+void test_efa_cq_readerr_happy(void **state)
+{
+	ssize_t rc;
+	struct efa_cq_readerr_state *s = calloc(1, sizeof(struct efa_cq_readerr_state));
+
+	assert_non_null(s);
+
+	rc = setup_readerr(s);
+	assert_int_equal(rc, 0);
+
+	s->efa_cq->ibv_cq_ex->wr_id = 0;
+	s->efa_cq->ibv_cq_ex->status = 1;
+
+	rc = efa_cq_readerr(&s->efa_cq->util_cq.cq_fid, s->entry, 0);
```

Comment:
Do you need `entry` in the state? would a local variable work for this case?

Reply:
We can. But I think it's encapsulated in the setup function that's shared with other test cases.

Reply:
I am not sure why it is better. To encapsulate it, you had to introduce a new struct.

Reply:
A new struct is a good thing. We can set up required variables, etc. in a single call to `setup_xxx`, and reuse it for different test cases, e.g. write once use everywhere.

Reply:
> A new struct is a good thing.

Not necessary.

> We can set up required variables, etc. in a single call to setup_xxx

But `entry` does not need to be setup.